### PR TITLE
Image.toByteData and Picture.toImage implementations (#2)

### PR DESCRIPTION
**Pull request by deakjahn**
## Description

* `Image.toByteData()` was not implemented in either DomCanvas or CanvasKit. This PR covers **both.**
* `Picture.toImage()` was not implemented in either DomCanvas or CanvasKit. This PR covers **CanvasKit** (I also have a solution for DomCanvas but that would need some discussion, it might clash with functionality somewhere else; details in a comment below).

## Related Issues

* https://github.com/flutter/flutter/issues/59072
* https://github.com/flutter/flutter/issues/42767
* https://github.com/flutter/flutter/issues/49857

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [ ] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by deakjahn** [comment]
@yjbanov Yegor, here I am with a new copy of the PR. This one now calls for the six changing files all right, so I hope for the best.

**Comment by deakjahn** [comment]
I copy this here, this is needed to discuss how to make this cover the DomCanvas case as well. If the PR itself lands, we could return to this:

## DomCanvas case

`EnginePicture.toImage()` has an implementation but it results in a bug. The reason for the bug is this fragment:

```dart
// lib/engine/engine/picture.dart

final BitmapCanvas canvas = BitmapCanvas(imageRect);
recordingCanvas.apply(canvas, imageRect);
final String imageDataUrl = canvas.toDataUrl();
```

when it comes to the third line, there will be no `_canvas` in the underlying `_CanvasPool` object, so calling `BitmapCanvas.toDataUrl()` will fail. Most canvas drawing operations make sure there is a canvas and context created, but not all of them do so. `recordingCanvas.apply()` will inevitably call `drawItemRect()` here that does not. So, the solution is to make sure that we do:

```dart
// lib/engine/engine/picture.dart

@override
Future<ui.Image> toImage(int width, int height) async {
  final ui.Rect imageRect = ui.Rect.fromLTRB(0, 0, width.toDouble(), height.toDouble());
  final BitmapCanvas canvas = BitmapCanvas(imageRect);
  canvas.forceCreate(); //<<<
  //canvas.drawPaint(SurfacePaintData()..color = ui.Color(0xFFFFFF)); //<<<
  recordingCanvas.apply(canvas, imageRect);
  final String imageDataUrl = canvas.toDataUrl();
  final html.ImageElement imageElement = html.ImageElement()
    ..src = imageDataUrl
    ..width = width
    ..height = height;
  return HtmlImage(
    imageElement,
    width,
    height,
  );
}
```

Either line marked with `//<<<`. Although the second might work, I'd suggest to create a specific function in `BitmapCanvas` that doesn't have any side effect and calling that:

```dart
// lib/engine/engine/bitmap_canvas.dart

/// Makes sure the underlying canvas is created.
void forceCreate() {
  final dummy = _canvasPool.contextHandle;
}
```

We're not done yet, the resulting image will be completely black. Unfortunately, the `drawItemRect()` mentioned earlier disposes of the canvas, so by the time we call `canvas.toDataUrl()`, it's already gone.

```dart
// lib/engine/engine/bitmap_canvas.dart

@override
void drawImageRect(
  ui.Image image, ui.Rect src, ui.Rect dst, SurfacePaintData paint) {

  ...

//    _closeCurrentCanvas(); //<<< this kills Picture.toImage()
}
```

This might mean some refactoring because, as the comments of `_closeCurrentCanvas()` reveal, this call *is* required in other scenarios but it's detrimental here. A possible solution would be to pass a boolean (with default value) to signal whether this last line should be called but being an overridden function, this requires modification in other parts of the code as well. Still, I don't think this can be avoided in this case.

**Comment by deakjahn** [comment]
Mostly done, and I don't think I can be held responsible legally for those font registration errors in those failing checks. ;-))

File: lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart [hunk 1]
```diff
@@ -1629,3 +1675,39 @@ external Object? get _finalizationRegistryConstructor;
 
 /// Whether the current browser supports `FinalizationRegistry`.
 bool browserSupportsFinalizationRegistry = _finalizationRegistryConstructor != null;
+
+class SkData {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1679]: This needs to be annotated with `@JS()`

Comment (deakjahn) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1680]: Sure. Used to be before the Big Reorganization. :-)

**Comment by deakjahn** [comment]
@yjbanov So, how shall we proceed? I'm hesitant to tick _All existing and new tests are passing_ because it isn't literally true, even if I suspect the failing tests show problems outside of the scope of this PR.

**Comment by yjbanov** [comment]
> @yjbanov So, how shall we proceed? I'm hesitant to tick _All existing and new tests are passing_ because it isn't literally true, even if I suspect the failing tests show problems outside of the scope of this PR.

Could you please rebase the PR on top of HEAD, and I'll take a look at the failures.

**Comment by deakjahn** [comment]
Gee, that killed it the last time...

**Comment by deakjahn** [comment]
@yjbanov Only one this time, apart from dear old Luci. That _Linux Web Engine_ is quite strange, I don't recall anything similar earlier. It seems calling the appropriate functions but blowing up somewhere deep inside the egine.

**Comment by yjbanov** [comment]
Trying to build this PR locally to see what's up.

**Comment by yjbanov** [comment]
The issue is that you need to pass `colorSpace` to `SkImageInfo`.

File: lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart [hunk 2]
```diff
@@ -1641,3 +1687,40 @@ external Object? get _finalizationRegistryConstructor;
...
+  external int get BytesSize64;
+  external SkColorSpace get ColorSpace;
+  external SkColorType get ColorType;
+  external int get Height;
+  external bool get IsEmpty;
+  external bool get IsOpaque;
+  external SkRect get Rect;
+  external int get RowBytes;
+  external int get RowBytes64;
+  external int get Width;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1714]: I think the field names above must exactly match the names in CanvasKit and in the factory below (for example "width" and "Width" is wrong; both should be "width").

File: lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart [hunk 3]
```diff
@@ -1641,3 +1687,40 @@ external Object? get _finalizationRegistryConstructor;
...
+  external factory SkImageInfo({
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1715]: nit: Flutter style prefers putting factories and constructors above everything else.

File: lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart (hunk 2 above)
Comment (deakjahn) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1714]: Oops, I have might inadvertently stumbled upon the documentation of the arch-enemy. :-)

File: lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart [hunk 4]
```diff
@@ -1705,14 +1705,14 @@ class SkImageInfo {
...
-  external SkRect get bounds();
-  external int get width();
+  external SkAlphaType alphaType();
+  external SkColorSpace colorSpace();
+  external SkColorType colorType();
+  external int height();
+  external bool isEmpty();
+  external bool isOpaque();
+  external SkRect bounds();
+  external int width();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1715]: These should be getters I imagine. However, getters don't need the `()`, for example `external bool get isEmpty;` (notice no `()` at the end).

Comment (deakjahn) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1715]: I don't know any more. The doc I finally found explicitely says they are functions, using ().

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1715]: Hmm. Interesting. According to the `@JS` bindings it's a plain object (in JS you'd write `{ width: x, height: y, aphaType: z, colorSpace: a, colorType: b }`). Perhaps there's a difference between `SkImageInfo` constructed by JS and one returned by CanvasKit?

Why don't you do what works now, and we can clean this up later?

/cc @kjlubick

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1715]: According to CanvasKit bindings it should be a POJO: https://github.com/google/skia/blob/1d4b92d87806e0d3ed8b4155da98d7d0ac757bd6/modules/canvaskit/canvaskit_bindings.cpp#L1873

That is, there shouldn't be any methods/functions on `SkImageInfo`, only properties.

File: lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  test('Convert Canvas to Picture', () async {
+    final SurfaceSceneBuilder builder = SurfaceSceneBuilder();
+    final Picture testPicture = await _drawTestPictureWithCircle(region);
+    builder.addPicture(Offset.zero, testPicture);
+
+    html.document.body.append(builder
+        .build()
+        .webOnlyRootElement);
+
+    await matchGoldenFile('canvas_to_picture.png', region: region, write: true);
```
Comment (yjbanov) [inline lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart:38]: For this to pass on CI you'll need to remove `write: true`. However, you will also need to add the new screenshot file to https://github.com/flutter/goldens (example PR: https://github.com/flutter/goldens/pull/103).

If that's too much ceremony, feel free to comment out this line in this PR, and I can do the dance with the screenshots in a follow up PR.

**Comment by deakjahn** [comment]
Same error with _Linux Web Engine_. I can't find any reference to color spaces in that log.

File: lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart (hunk 5 above)
Comment (deakjahn) [reply lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart:38]: I simply can't get the screenshot. The test doesn't run on either Windows or Linux for me locally (on Windows by design, on Linux it should work but it doesn't, Chrome starts and then nothing ever happens).

File: lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart (hunk 4 above)
Comment (kjlubick) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart:1715]: That is correct, the SkImageInfo is just a POJO

Comment (deakjahn) [reply lib/web_ui/lib/src/engine/compositor/canvaskit_api.dart]: OK, I do, I do.

File: lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart (hunk 5 above)
Comment (deakjahn) [reply lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart]: Commenting out renders the test inactive. If it's OK like this, let's do so.

Comment (yjbanov) [reply lib/web_ui/test/golden_tests/engine/canvas_to_picture_test.dart]: Filed https://github.com/flutter/flutter/issues/63711 to follow up later.

**Comment by yjbanov** [comment]
@deakjahn Sorry, I renamed the `compositor` directory to `canvaskit` today, so this PR needs to be rebased again 😅 

It looks like this PR is almost good to go. The only remaining issue is an unused import.

**Comment by deakjahn** [comment]
An import? Did I add it? Which one? I don't remember needing to add any import at all but that doesn't mean a thing. :-)

**Comment by deakjahn** [comment]
I have a suggestion that might help us. It's only six files, after all. So, if

* I close this PR, too,
* Delete my fork,
* For the engine again,
* Copy the six modified files and commit them in one go,
* Create a #3 PR

that might fly, I guess.

**Comment by yjbanov** [comment]
That plan SGTM. I'm wondering if Github allows downloading a PR as a patch file that could be applied locally as a single commit 🤔

**Comment by deakjahn** [comment]
It would if I squashed it into one. But then there would be no need...

**Pull request closed without merging**